### ISSUE TITLE
Hope to support the following special ASCII character input, thank you!

### ISSUE THREAD

**Issue by Helaer** [Genymobile/scrcpy#2816]
The following is the special ASCII character input supported by Qtscrcpy. I hope that scrcpy can also support the following special ASCII character input. I am not good at C language, so come to ask for help, thank you!

```diff
@@ -279,6 +279,7 @@ AndroidKeycode InputConvertNormal::convertKeyCode(int key, Qt::KeyboardModifiers
        keyCode = AKEYCODE_0;
        break;
    case Qt::Key_1:
    case Qt::Key_Exclam:// !
        keyCode = AKEYCODE_1;
        break;
    case Qt::Key_2:
@@ -288,15 +289,19 @@ AndroidKeycode InputConvertNormal::convertKeyCode(int key, Qt::KeyboardModifiers
        keyCode = AKEYCODE_3;
        break;
    case Qt::Key_4:
    case Qt::Key_Dollar://$
        keyCode = AKEYCODE_4;
        break;
    case Qt::Key_5:
    case Qt::Key_Percent:// %
        keyCode = AKEYCODE_5;
        break;
    case Qt::Key_6:
    case Qt::Key_AsciiCircum: //^
        keyCode = AKEYCODE_6;
        break;
    case Qt::Key_7:
    case Qt::Key_Ampersand: //&
        keyCode = AKEYCODE_7;
        break;
    case Qt::Key_8:
@@ -308,6 +313,67 @@ AndroidKeycode InputConvertNormal::convertKeyCode(int key, Qt::KeyboardModifiers
    case Qt::Key_Space:
        keyCode = AKEYCODE_SPACE;
        break;
    case Qt::Key_Comma://,
    case Qt::Key_Less://<
        keyCode = AKEYCODE_COMMA;
        break;
    case Qt::Key_Period://.
    case Qt::Key_Greater://>
        keyCode = AKEYCODE_PERIOD;
        break;
    case Qt::Key_Minus://-
    case Qt::Key_Underscore: //_
        keyCode = AKEYCODE_MINUS;
        break;
    case Qt::Key_Equal://=
        keyCode = AKEYCODE_EQUALS;
        break;
    case Qt::Key_BracketLeft://[
    case Qt::Key_BraceLeft: //{
        keyCode = AKEYCODE_LEFT_BRACKET;
        break;
    case Qt::Key_BracketRight://]
    case Qt::Key_BraceRight: //}
        keyCode = AKEYCODE_RIGHT_BRACKET;
        break;
    case Qt::Key_Backslash:// \ ????
    case Qt::Key_Bar: //|
        keyCode = AKEYCODE_BACKSLASH;
        break;
    case Qt::Key_Semicolon://;
    case Qt::Key_Colon: //:
        keyCode = AKEYCODE_SEMICOLON;
        break;
    case Qt::Key_Apostrophe://'
    case Qt::Key_QuoteDbl: //"
        keyCode = AKEYCODE_APOSTROPHE;
        break;
    case Qt::Key_Slash:// /
    case Qt::Key_Question://?
        keyCode = AKEYCODE_SLASH;
        break;
    case Qt::Key_At://@
        keyCode = AKEYCODE_AT;
        break;
    case Qt::Key_Plus://+
        keyCode = AKEYCODE_PLUS;
        break;
    case Qt::Key_QuoteLeft://`
    case Qt::Key_AsciiTilde://~
        keyCode = AKEYCODE_GRAVE;
        break;
    case Qt::Key_NumberSign: //#
        keyCode = AKEYCODE_POUND;
        break;
    case Qt::Key_ParenLeft: //(
        keyCode = AKEYCODE_NUMPAD_LEFT_PAREN;
        break;
    case Qt::Key_ParenRight: //)
        keyCode = AKEYCODE_NUMPAD_RIGHT_PAREN;
        break;
    case Qt::Key_Asterisk: //*
        keyCode = AKEYCODE_STAR;
        break;
    }
    return keyCode;
}
```

**Comment by rom1v** [comment]
These keys are handled as text events (see [here](https://blog.rom1v.com/2018/03/introducing-scrcpy/#handle-text-input)). Their character differ depending on the charmap/language.

Why do you need them as key events?

**Comment by Helaer** [comment]
> 



> 这些键作为文本事件处理（请参阅[此处](https://blog.rom1v.com/2018/03/introducing-scrcpy/#handle-text-input)）。他们的性格因魅力图/语言而异。
> 
> 为什么需要它们作为关键事件？



> These keys are handled as text events (see [here](https://blog.rom1v.com/2018/03/introducing-scrcpy/#handle-text-input)). Their character differ depending on the charmap/language.
> 
> Why do you need them as key events?

In Chinese input, Chinese punctuation cannot be input normally. Qtscrcpy is optimized in this respect. It can input Chinese punctuation normally when inputting Chinese. This is very bad for me who often input text. I hope you can make some optimizations. Thank you!

**Comment by rom1v** [comment]
Could you test this (very old) commit: https://github.com/Genymobile/scrcpy/issues/87#issuecomment-413292204

**Comment by Helaer** [comment]
> 你能测试这个（很老的）分支吗：[`rawkeyevents`](https://github.com/Genymobile/scrcpy/commits/rawkeyevents)（[#87（评论）](https://github.com/Genymobile/scrcpy/issues/87#issuecomment-413292204)）

I downloaded the 1.3 branch, but the -k option does not work.

![image](https://user-images.githubusercontent.com/52807664/142850955-84f75722-8da1-4ce2-8601-87249805e1bf.png)

**Comment by rom1v** [comment]
There is no binary for this `rawkeyevents` branch. You have to build it manually.

**Comment by Helaer** [comment]
> 

Sorry, I don’t know how to build a binary file, can you help me build it, thank you!

**Comment by Helaer** [comment]
> 你能测试这个（很老的）分支吗：（[#87（评论）](https://github.com/Genymobile/scrcpy/issues/87#issuecomment-413292204)）[`rawkeyevents`](https://github.com/Genymobile/scrcpy/commits/rawkeyevents)[](https://github.com/Genymobile/scrcpy/issues/87#issuecomment-413292204)

I just found a friend to help me compile, the test result is that the above special ASCII characters can be input, but this branch uses Ctrl+K to switch, this experience is not very good, I hope to add --raw-key-events Or -k option to start by default.
![image](https://user-images.githubusercontent.com/52807664/142861095-cc379f4c-aa1d-481e-b4e5-20785be8cd39.png)

**Comment by Helaer** [comment]
I hope the 1.12.1 version can add the --raw-key-events or -k function, so that the text input of scrcpy can be perfectly inputted with the mobile input method. If you can do this, it will be great!

**Comment by Helaer** [comment]
@rom1v  Can the --raw-key-events or -k option be added in the next version? I think this feature is what most people need, thank you!

**Comment by rom1v** [comment]
Probably something similar, yes.

If you find a way to configure the physical keyboard layout even when a physical keyboard (or HID) is not connected, that would be great :wink:

**Comment by Helaer** [comment]
> 

@rom1v Hello, I have many friends who need this function, Can you add --raw-key-events to the next version? and I hope you can release the next version as soon as possible, Thank you.

**Comment by rom1v** [comment]
> Can you add --raw-key-events to the next version?

I just answered in the previous comment: yes, I'll implement it.

**Comment by Helaer** [comment]
> > Can you add --raw-key-events to the next version?
> 
> I just answered in the previous comment: yes, I'll implement it.
> 
> > I hope you can release the next version as soon as possible
> 
> Depends on my free time, I also have other things to implement before a new release.

Okay, thank you for your contribution.

**Comment by rom1v** [comment]
> Oh, the reason why scrcpy uses a mix of key events and text events is that key events depend on the physical device keyboard layout (so most of the time, it does not behave like expected).
>
> But when an [HID keyboard](https://github.com/Genymobile/scrcpy#physical-keyboard-simulation-hid) is enabled (or a real physical keyboard is plugged), the layout may be configured in Settings → System → Languages and input → [Physical keyboard](https://github.com/Genymobile/scrcpy/pull/2632#issuecomment-923756915).
>
> This option disappears from the settings as soon as the physical keyboard is disconnected, but its configuration remains for injecting key events! :tada:

My test was incorrect, this is totally false. Physical keyboard layout configuration has no impact on raw key events.

Anyway, I implemented the option: #2831.
